kubernetes rack awareness (#904)
redpanda-data/documentation: docs/platform/data-management/kubernetes-rack-awareness.mdx

---

> title: Kubernetes Helm chart: rack awareness

Author: :+1:

---

<head>
    <meta name="title" content="Kubernetes Helm chart: rack awareness | Redpanda Docs"/>
    <meta name="description" content="Setting up rack awareness with the Kubernetes helm chart."/>
</head>

> import styles from './rack-awareness.module.css';

Reviewer: ```suggestion
```

No need to import this here. The original 'Rack Awareness' page customizes the width of a table using these styles.

> For more information about what Redpanda rack awareness is and why it's used, see [data-management/rack-awareness](./rack-awareness)

Reviewer 2: The introduction needs more context. How about:
```suggestion
To use rack awareness with Kubernetes, you must configure it in your `redpanda.yaml` file, then apply the change to your cluster. For more information about what Redpanda rack awareness is and why it's used, see [data-management/rack-awareness](./rack-awareness).
```

Author: Except helm chart users don't have to do that. The chart handles that for them.

Reviewer: ```suggestion
When you enable rack awareness in the Helm chart, Kubernetes failure zones are treated as racks. Redpanda maps each rack to a failure zone and places partition replicas across them. For more information about what Redpanda rack awareness is and why it's used, see [data-management/rack-awareness](./rack-awareness). For more details about Kubernetes failure zones, see the [Kubernetes documentation](https://kubernetes.io/docs/setup/best-practices/multiple-zones/).
```

## Setting up rack awareness with the Redpanda Helm chart

By convention, the major cloud services annotate Kubernetes Nodes with the availability zone that the Node instance is placed in.
To use that annotation to inform Redpanda to use those zones for fault tolerance at the data level, apply values to enable the feature:

> ### Step 1

Author: :+1:

> Create a configuration file to enable rack awareness:

Author: This is the convention we've been using in the TLS and SASL docs. The helm command can reference multiple values files and will merge them with the defaults (and any existing configurations in the case of upgrades). When we get to use-case documenting, I plan on making the use of multiple values files more obvious.

For now I think we can count on users being sophisticated enough to know that the file is an example and either combining the settings they need into a single yaml file, or knowing enough about helm to use multiple `--values` flags.

_rack_awareness.yaml_

```yaml
rackAwareness:
    enabled: true
serviceAccount:
    create: true
rbac:
    enabled: true
```

> See the Helm chart's [values.yaml](https://github.com/joejulian/helm-charts/tree/main/charts/redpanda/values.yaml) to see additional options to rackAwareness, serviceAccount, and rbac.

Author: :+1:

> See the Helm chart's [values.yaml](https://github.com/joejulian/helm-charts/tree/main/charts/redpanda/values.yaml) to see additional options to rackAwareness, serviceAccount, and rbac.

Reviewer 2: Thanks, Jake. I noticed that, too!

> See the Helm chart's [values.yaml](https://github.com/joejulian/helm-charts/tree/main/charts/redpanda/values.yaml) to see additional options to rackAwareness, serviceAccount, and rbac.

Author: ```suggestion
See the Helm chart's [values.yaml](https://github.com/redpanda-data/helm-charts/tree/main/charts/redpanda/values.yaml) for additional options to `rackAwareness`, `serviceAccount`, and `rbac`
```

> The rackAwareness configuration defaults to using the [well-known](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone) Node annotation (`rackAwareness.nodeAnnotation`): `topology.kubernetes.io/zone`, setting the redpanda configuration flag to enable rack awareness, and setting the `rack` configuration to the value of the node annotation.

Reviewer 2: This looks like a list of three default values. Could you rearrange them into a bullet list? This is my attempt at a rewrite (I don't understand the first bullet, so I need your help):

```suggestion
The `rackAwareness` configuration uses the following default values:
- The [well-known](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone) node annotation (`rackAwareness.nodeAnnotation`): `topology.kubernetes.io/zone`.
- The `redpanda.yaml` configuration flag is set to enable rack awareness.
- The `rack` configuration is set to the value of the node annotation.
```

Author: sure

The serviceAccount configuration allows the Helm chart to create a ServiceAccount.

> The rbac configuration allows the Helm chart to create ClusterRoles and ClusterRoleBindings that allow the Redpanda initialization container to read the annotation from the Node.

Reviewer 2: Here's an idea:
What if you put these explanations as comment text in the `.yaml` file example? That way, the explanations will be provided right away, instead of having them several lines later (and losing the thread).

Reviewer: Could we add an explanation for why the ServiceAccount is needed?

> ### Step 2

Author: Sounds good to me. I was just trying to follow the style guide. I never liked the "Step N" thing anyway.

```suggestion
### Install or Update your Helm release
```

> Create a new Redpanda cluster or apply this change to an existing Redpanda cluster with Helm:

Author: This actually could create it from scratch. I thought it would be valuable as a user that's interested in this topic could create their own cluster with rack awareness enabled just using this one page of copy/paste blocks, a great way to see how it works without having to cross reference other pages.

```sh
helm repo add redpanda https://charts.redpanda.com/
helm repo update
helm upgrade --install redpanda redpanda/redpanda redpanda \
  --namespace redpanda --create-namespace \
  --values rack_awareness.yaml
```

## Annotating nodes in non-cloud settings

> The common convention for annotating Kubernetes Nodes is to use the `topology.kubernetes.io/zone` key. Label the nodes based on your own fault-tolerance requirements. Set this annotation by running:

Reviewer 2: How about the answer to question 2?

> The common convention for annotating Kubernetes Nodes is to use the `topology.kubernetes.io/zone` key. Label the nodes based on your own fault-tolerance requirements. Set this annotation by running:

Author: I don't have an answer for #2. There are at least dozens of ways of building infrastructure. Maybe racks, maybe sections of rack, maybe "modules", maybe rows, maybe they're running multiple VMs on physical machines and they want to ensure they're on different machines. My expectation is that if a user is not using a cloud infrastructure, whomever builds their infra knows the answer to what that requirement is.

If we need to explain this, it would probably be better done out of band (footnote?)

```sh
kubectl annotate node dc1r12c22 topology.kubernetes.io/zone=dc1r12
```

> Where dc1r12c22 is an example Node name of compute node 22 in rack 12 in datacenter 1. We set the annotation to dc1r12 so that any Redpanda pod that gets scheduled on that Node with have it's `rack` setting configured to `dc1r12`.

Author: :+1:

---
title: Configuring Rack Awareness in Kubernetes using the Helm Chart
---

<head>
    <meta name="title" content="Kubernetes Helm chart: rack awareness | Redpanda Docs"/>
    <meta name="description" content="Setting up rack awareness with the Kubernetes helm chart."/>
</head>

For more information about what Redpanda rack awareness is and why it's used, see [data-management/rack-awareness](./rack-awareness)

## Setting up rack awareness with the Redpanda Helm chart

> By convention, the major cloud services annotate Kubernetes Nodes with the availability zone that the Node instance is placed in.

Reviewer: The Kubernetes docs seem to prefer failure zones to availability zones.

To use that annotation to inform Redpanda to use those zones for fault tolerance at the data level, apply values to enable the feature:

### Create a configuration file

Create a configuration file to enable rack awareness:

_rack_awareness.yaml_

```yaml
rackAwareness:
    enabled: true
serviceAccount:
    create: true
rbac:
    enabled: true
```

See the Helm chart's [values.yaml](https://github.com/redpanda-data/helm-charts/tree/main/charts/redpanda/values.yaml) for additional options to `rackAwareness`, `serviceAccount`, and `rbac`

The `rackAwareness` configuration uses the following default values:
- The [well-known](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone) node annotation (`rackAwareness.nodeAnnotation`): `topology.kubernetes.io/zone`.
- The `redpanda.yaml` configuration flag is set to enable rack awareness.
- The `rack` configuration is set to the value of the node annotation.

> 

Author: This would require the user to make a decision while trying to follow copy/paste examples. I'd prefer we defer choices that would require decisions until the user has seen how it's expected to work.

> 

Reviewer: What happens if the user doesn't use `topology.kubernetes.io/zone`? Will rack awareness just not work? Or will the chart notify the user somehow?

I think it's best to prevent such an experience. Maybe we can provide two paths in the doc.

> 

Author: Sure, whatever you think!

> The serviceAccount configuration allows the Helm chart to create a ServiceAccount.

Reviewer: ```suggestion
The `serviceAccount` configuration allows the Helm chart to create a ServiceAccount.
```

> The rbac configuration allows the Helm chart to create ClusterRoles and ClusterRoleBindings that allow the Redpanda initialization container to read the annotation from the Node.

Reviewer: ```suggestion
The `rbac` configuration allows the Helm chart to create ClusterRoles and ClusterRoleBindings that allow the Redpanda initialization container to read the annotation from the Node.
```

> The `rbac` configuration allows the Helm chart to create ClusterRoles and ClusterRoleBindings that allow the Redpanda initialization container to read the annotation from the Node.

Author: per https://github.com/redpanda-data/documentation/pull/904#discussion_r1046140674 which GitHub seems to have lost.

The ServiceAccount, combined with the ClusterRole and ClusterRoleBinding (so far) allow the pod to have read access to get and list the Nodes in order to retrieve the annotation. In the future, they might do more. I'm hesitant to document that here in case future needs come up. Saying that we're only doing one security-related thing and doing more could be a breach of trust.

### Install or Update your Helm release

Create a new Redpanda cluster or apply this change to an existing Redpanda cluster with Helm:

```sh
helm repo add redpanda https://charts.redpanda.com/
helm repo update
helm upgrade --install redpanda redpanda/redpanda redpanda \
  --namespace redpanda --create-namespace \
  --values rack_awareness.yaml
```

> ## Annotating nodes in non-cloud settings

Reviewer: This part seems like an afterthought. I think we rather have a prerequisite:

- Make sure that each of your Kubernetes Nodes is annotated with a key that corresponds to a failure zone. The Helm chart uses the `topology.kubernetes.io/zone` annotation by default.

If you are using a cloud platform, the major cloud providers annotate Kubernetes Nodes with the availability zone in which the Node instance is placed. For example `topology.kubernetes.io/zone=us-east-1a`.

If you are not using a cloud platform, annotate your Nodes based on your fault-tolerance requirements. For example:

```bash
kubectl annotate node dc1r12c22 topology.kubernetes.io/zone=dc1r12
```

In this example, dc1r12c22 is the node name for compute node 22 in rack 12 in data center 1. The annotation is set to dc1r12 so that any Redpanda pod that gets scheduled on that Node will have its `rack` configuration set to `dc1r12`.

Author: Do you want to set that up with Tabs?

Reviewer: It would only be a short couple of sentences under the heading 'Prerequisites'.

It's equally annoying getting to the end of a document and realizing that your cluster wasn't set up properly to use the feature in the first place.

Author: Whatever you think is best. I was just explaining my thought process. I'm not a docs expert. :)
